library.go

```diff
@@ -837,34 +837,99 @@ func range_(thread *Thread, fn *Builtin, args Tuple, kwargs []Tuple) (Value, err
 	if err := UnpackPositionalArgs("range", args, kwargs, 1, &start, &stop, &step); err != nil {
 		return nil, err
 	}
-	list := new(List)
+
+	// TODO(adonovan): analyze overflow/underflows cases for 32-bit implementations.
+
+	var n int
 	switch len(args) {
 	case 1:
 		// range(stop)
 		start, stop = 0, start
 		fallthrough
 	case 2:
 		// range(start, stop)
-		for i := start; i < stop; i += step {
-			list.elems = append(list.elems, MakeInt(i))
+		if stop > start {
+			n = stop - start
 		}
 	case 3:
 		// range(start, stop, step)
-		if step == 0 {
-			return nil, fmt.Errorf("range: step argument must not be zero")
-		}
-		if step > 0 {
-			for i := start; i < stop; i += step {
-				list.elems = append(list.elems, MakeInt(i))
+		switch {
+		case step > 0:
+			if stop > start {
+				n = (stop-1-start)/step + 1
 			}
-		} else {
-			for i := start; i >= stop; i += step {
-				list.elems = append(list.elems, MakeInt(i))
+		case step < 0:
+			if start > stop {
+				n = (start-1-stop)/-step + 1
 			}
+		default:
+			return nil, fmt.Errorf("range: step argument must not be zero")
 		}
 	}
-	return list, nil
+
+	return rangeValue{start: start, stop: stop, step: step, len: n}, nil
+}
+
+// A rangeValue is a comparable, immutable, indexable sequence of integers
+// defined by the three parameters to a range(...) call.
+// Invariant: step != 0.
+type rangeValue struct{ start, stop, step, len int }
+
+var (
+	_ Indexable  = rangeValue{}
+	_ Sequence   = rangeValue{}
+	_ Comparable = rangeValue{}
+)
+
+func (r rangeValue) Len() int          { return r.len }
+func (r rangeValue) Index(i int) Value { return MakeInt(r.start + i*r.step) }
```

Comment:
Should there be a bounds check here, or are callers responsible for calling `Len` first?